keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+GET operations for those CRDs, dynamic informers will have to be used that get
+started and stopped as needed.
+
+### Risks and Mitigations
+
+Granting kube-scheduler permission to read arbitrary types could be used by an
+attacker to extract information. However, for that an attacker first would have
+to gain control of the scheduler, get it to execute API calls, and then
+retrieve the result. This seems like a low risk.
+
+Using a single NodeResourceCapacity object per node is simple and convenient
+also for admins, but might run into size limitations. This should not be a
+problem in practice because the number of drivers per node is expected to be
+low and the capacity information from each should be small. Limits for the
+JSON fragment can be introduced to prevent excessive sizes.
+
+## Design Details
+
+### Decoding claim parameters
+
+Extracting the numeric parameters can be done without knowing the CRD:
+
+- Look up the resource based on group and kind with a REST mapper.
+
+- Retrieve the object with a `dynamic.Client` as an `unstructured.Unstructured`
+  instance (basically a nested map).
+
+- Retrieve the sub-map via the JSON path.
+
+- Re-encode that part as JSON with the normal `encode/json` package.  In the
+  example above, this will include `kind` and `apiVersion:
+  counter.dra.config.k8s.io` because those are the defaults and only allowed
+  values.
+
+- Decode into the numeric parameter with the apimachinery JSON decoder,
+  ignoring unknown fields like `env` in this example.
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables numeric parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via numeric parameters, then NumericParameters must list all claim
+    // parameter types that are allowed for claims.
+    NumericParameters []NumericParameterType
+}
+
+type NumericParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// FieldPath is a dot separarated JSON path that defines where in the
+	// parameter object the numeric parameters are embedded.
+	FieldPath string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResourceCapacity
+
+For each node, a NodeResourceCapacity objects gets created:
+
+```
+// NodeResourceCapacity gets published by kubelet. Its name
+// matches the name of the node and the node is the owner.
+//
+// Capacity may get added, but should not get removed because
+// it would make scheduling decisions based on the old capacity
+// invalid.
```

Comment:
What if a device goes unhealthy? How does one find out that a certain device is no longer available for scheduling if we cannot "decrease" the capacity here.

In the traditional device plugin API, this is accomplished by keeping a persistent watch open, where the plugin updates the kubelet with a full list of devices anytime that list needs to change (e.g. a device goes unhealthy and is no longer part of the list, or a new device is hot-plugged and gets added to the list).

Reply:
The mechanism would be the same here, and in the case of an "unhealthy" device the capacity would get decreased. I'll remove this comment here and instead elaborate on this in a section of the KEP.